fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -2289,10 +2352,87 @@ public Builder allowUniquePendingState(boolean allow) {
+             *  * "Single index by records" may continue a mutually built session.
+             * @param allow if true, allow takeover.
+             * @return this builder
+             */
+            public Builder allowTakeoverContinue(boolean allow) {
+                this.doAllowTakeoverContinue = allow;
+                return this;
+            }
+
+            /**
+             * Call {@link #setMutualIndexing(boolean)} with default true.
+             * @return this builder
+             */
+            @API(API.Status.EXPERIMENTAL)
+            public Builder setMutualIndexing() {
+                this.useMutualIndexing = true;
+                return this;
+            }
+
+            /**
+             * If set, allow mutual (parallel) indexing. In this state, the indexer will assume that other indexers, called
+             * by other threads/processes/systems with the exact same parameters, are attempting to concurrently build this
+             * index. To allow that, the indexer will:
+             *   1. Use an a-synchronized session (i.e. avoid the indexing lock)
+             *   2. Divide the records space to fragments, then iterate the fragments in a minimal interference way, while
+             *      indexing each fragment independently.
+             *   3. Handle indexing conflicts, when occurred.
+             * The caller may use any number of concurrent indexers as desired. By default, the fragments are
+             * split by primary key boundaries (the boundaries can also be preset by {@link #setMutualIndexingBoundaries(List)}).
+             *
+             * @param useMutualIndexing if true, allow this state.
+             * @return this builder
+             */
+            @API(API.Status.EXPERIMENTAL)
+            public Builder setMutualIndexing(final boolean useMutualIndexing) {
+                this.useMutualIndexing = useMutualIndexing;
+                if (!useMutualIndexing) {
+                    useMutualIndexingBoundaries = null;
+                }
+                return this;
+            }
+
+            /**
+             * Same as {@link #setMutualIndexing()}, but will use a pre-defined set of keys to split
+             * the records space to fragments. {@code null} can be used to clear this value.
+             * To ensure defining the full records range, the boundaries list can begin and end with a `null`.
+             * @param primaryKeysBoundaries set of primary keys that will be used to split the records space to fragments.
+             * @return this builder
+             */
+            @API(API.Status.EXPERIMENTAL)
+            public Builder setMutualIndexingBoundaries(final List<Tuple> primaryKeysBoundaries) {
+                if (primaryKeysBoundaries == null || primaryKeysBoundaries.isEmpty()) {
+                    this.useMutualIndexingBoundaries = null;
+                } else {
+                    this.useMutualIndexingBoundaries = new ArrayList<>(primaryKeysBoundaries);
```

Comment:
Okay, this is a bit confusing, but it seems like either:

1. `.setMutualIndexBoundaries` and `.setMutualIndexing` should _only_ affect the boundaries list and the boolean flag respectively
2. `.setMutualIndexBoundaries` should enable `.setMutualIndexing` if given a non-null list and `.setMutualIndexing` should clear the boundaries list if set to false. If `.setMutualIndexBoundaries` is given `null`, then it should _not_ change `.setMutualIndexing`, because a `null` list is a valid set of boundaries, which I think was the crux of the issue during an earlier revision

I think my preference is for (1), the idea being that putting logic in a builder other than setting/clearing fields is a mistake, but if you want (2), I think the idea that the builder can maintain the invariant that `!useMutualIndexing` implies `useMutualIndexingBoundaries == null` (and thus also the contrapositive, `useMutualIndexingBoundaries != null` implies `useMutualIndexing`) is a manageable contract.

Reply:
With current version, (1) is chosen - I can surely see why you prefer it as it removes dependency of the caller's setup order and  to multiple, conflicting calls of the setting functions. 
During the final Builder's build, however, I assumed that if the caller had explicitly set a boundaries list, he doesn't expect them to be ignored in a non-mutual indexing. Does it make sense to you?

Reply:
Hm, I see. I think I was a bit thrown because I wasn't expecting `.build()` to change `useMutualIndexing`, which I think means that from an API perspective, it's actually closer to version (2). (In that if `mutualIndexingBoundaries != null`, then `useMutualIndexing` is also true by the time the object is actually built.)

I _think_ my preference between setting `useMutualIndexing` in `.build()` and setting it in `.setMutualIndexingBoundaries` is to prefer `.setMutualIndexingBoundaries` because (1) putting it there makes it a bit clearer to the user what state transitions such a call causes and (2) the code/state transitions between `.setMutualIndexing(false)` and `.setMutualIndexingBoundaries(nonEmptyList)` are more parallel (and I don't think there's a way to consolidate everything in `.build()` because `useMutualIndexing=false` and `mutualIndexingBoundaries!=null`).

I can see why you don't want to ignore the user if they set an explicit boundary list, I suppose. There's maybe a case where someone might want to have code that's something like:

```java
builder
    .setMutualIndexing(SystemConfig.shouldUseMutualIndexing())
    .setMutualIndexingBoundaries(SystemConfig.getMutualIndexingBoundaries())
```

And they want to seed their system configuration with a list of boundaries, but only actually use mutual indexing if the other configuration Boolean is set. But that's maybe a little far-fetched given that that implies that the user has a default set of mutual indexing boundaries that they know should work.

Reply:
Mmm... so what's the bottom line? 
The advantage of keeping it in the builder's build is for cases of, lets say, automatically initiating the builder with a boundaries list, then calling `setMutualIndexingBoundaries(null)` to clear it for a non-mutual indexers. That will generate an unsolicited meaning of the builder's history.
I have no idea how likely is this scenario, or any other.

Reply:
Cases like that are part of the reason why my initial suggestion was to make the builders `.setX` methods only set X and not try and set related values as well, even if it means that the user has to set two related parameters. At this point, I think what we have now is probably fine, though arguably, the more important thing is that the behavior is documented